gnetlist/docs/mk_verilog_syms.1
.TH mk_verilog_syms 1 "September 25th, 2013" "gEDA Project" 1.8.2.20130925
.SH NAME
mk_verilog_syms \- Create gate symbols for the gEDA Verilog netlister

.SH SYNOPSIS
.B mk_verilog_syms

.SH DESCRIPTION
.B mk_verilog_syms
creates a set of n-input gate symbols for the gEDA Verilog
netlister. It creates the following symbols:
.IP \(bu
and[2-9]
.IP \(bu
nand[2-9]
.IP \(bu
nor[2-9]
.IP \(bu
or[2-9]
.IP \(bu
xor[2-9]
.IP \(bu
xnor[2-9]

.SH AUTHOR
Mike Jarabek
.SH SEE ALSO
.BR gschem (1),
.BR gnetlist (1)
.SH COPYRIGHT
.nf
Copyright \(co 1999-2011 gEDA Contributors.  License GPLv2+: GNU GPL
version 2 or later.  Please see the `COPYING' file included with this
program for full details.
.PP
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.


NEWS
==============================================
 'gschem and Friends' Electronic Design Suite
==============================================

Copyright (C) 1998-2014 gEDA Developers

This file documents important user-visible changes in gEDA/gaf.  For
more information, please consult the `ChangeLog' file.

Notable changes in gEDA/gaf 1.9.2
=================================
* Visibility and attributes of several VHDL symbols have been fixed.

* Two obsolete scripts verilog2vhdl.sh and script.sed have been
  removed.

* Undocumented and ambiguous feature of the `spice-sdb' backend for
  gnetlist consisting in the possibility of using the "value="
  attribute instead of "file=" only for include blocks and only in
  embedding mode has been removed.

Notable changes in gEDA/gaf 1.9.1
=================================

* Unbalanced overbar markers are now rendered correctly.

* Most tools now no longer set the GEDADATA or GEDADATARC environment
  variables (they still obey them though).

* `gschem' now provides a path drawing tool, accessed using <A H> or
  the "Add Path" menu item.

  - Add a cusp node by left-clicking.

  - Add a Bezier curve node by left-dragging from the node point to
    the control point.

  - End the path by clicking on the same node point twice, or clicking
    on the start node to close the path.

* On desktops that reparent menu bars (e.g. Ubuntu Unity) it is once
  again possible to paste the clipboard using the menus.

* A number of `gschem' editing behaviour changes were made:

  - Double-clicking on an entry in the "In Use" tab of the component
    selector now behaves the same way as double-clicking in the
    "Libraries" tab.

  - When new attributes are attached to a net, they are placed at a 50
    unit offset.

  - The text color is now used as the default color for `pinlabel='
    attributes.

  - Pin attributes are no longer automatically replaced when rotating
    or mirroring pins.

  - It is possible to edit a text object even when other non-text
    objects are selected.

  - Control grips are now always drawn at the same on-screen size.

* The multi-attribute editor dialog in `gschem' has been extended to
  support more cases:

  - It now supports editing attributes of multiple selected
    objects. These are aggregated into the dialog's attribute list.

    Multiple attributes with the same name attached to any given
    object are still shown in separate rows, but identically named
    attributes from multiple selected objects will share a row. If a
    row has differing values between objects, its value shows the text
    "<various>" in a muted colour.

  - Where an attribute is present in some, but not all selected
    objects its name in the list is shown in red. A new popup menu
    option allows copying such attributes to all selected objects.

  - If no complex, placeholder, pin, net or bus objects are selected,
    the dialog now falls back to editing any selected attribute text.

    This allows editing of unattached or floating attributes in
    symbols, or any attributes seleccted independantly of their owning
    object. In this fallback mode, all attributes with the same name
    shown in separate rows, and the "add attribute" functionality is
    disabled.

* `gschem' has new icons for many editing commands.

* Several bugs were fixed in `gnetlist':

  - Invalid attributes are ignored.

  - Correct detection of graphical symbols.

  - Components with no `refdes=' attribute are reported.

  - Crash fixes for `drc' and `switcap' backends.

* `gaf export' now accepts a `--scale' option.

* The `--size', `--margins' and `--align' options to `gaf export' can
  now accept their multiple arguments separated with a space ` ', a
  colon `:' or a semicolon `;'.

* `gaf config' now handles the `-p' option properly.

* `gnetlist', `gsymcheck' and `gaf' now output internationalised
  messages.

* New Scheme procedures were added. Please see the "geda-scheme" info
  manual for more details.

  - The (gschem action) module contains procedures for working with
    user editing actions (i.e. commands accessed via the menus or
    keybindings).

  - Two new hooks were added: `bind-keys-hook' is called whenever
    keymap bindings are changed; `action-property-hook' is called when
    action properties are changed.


Notable changes in gEDA/gaf 1.9.0
=================================

* A new schematic renderer library, `libgedacairo' has been added, and
  all tools have been updated to use it.  Related changes:

  - The following configuration functions have been removed:

         output-capstyle         paper-sizes
         output-color            postscript-prolog
         output-orientation      print-command
         output-type             setpagedevice-orientation
         paper-size              setpagedevice-pagesize

  - The following `gschem' configuration items have been added:

         print-paper             print-color
         print-orientation

  - `gschem' now generates PDF files rather than PostScript or EPS
    files.

  - `gschem' now uses a standard GTK print dialog (with printer
    selection, print preview, etc.)

* A new `gaf' command-line utility has been added.  It provides a
  number of subcommands:

  - `gaf export' can be used to convert schematic files to PNG, SVG,
    PDF, PS and EPS, with detailed control of paper size and margins.
    It supports multi-page PDF and PS output.

  - `gaf config' can be used to manipulate the project, user, and
    system configuration stores (see below).

  - `gaf shell' provides a Scheme REPL for batch manipulation of
    schematic and symbol files.

* New project, user and system configuration stores have been added,
  using a format similar to .desktop files.  They can be manipulated
  using the `gaf config' tool.  Configuration parameters for most
  tools will be converted to use this mechanism in future releases.

* The changes to highlighting of dangling net ends in gEDA/gaf 1.8.0

gnetlist/utils/mk_verilog_syms.c
/*  mk_verilog_syms.c: 
 *
 *  Create a set of n-input gate symbols for the geda verilog netlister.
 *
 *  symbols created:  and[2-9], nand[2-9], nor[2-9], or[2-9], nor[2-9],
 *  xor[2-9], xnor[2-9].
 *
 * Copyright (C) 1999-2010 Mike Jarabek
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02111-1301 USA.
 *
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <stdio.h>
#include <math.h>

/* local defines */

/* gEDA colour, these should probably come from somewhere else */
#define BLACK		0
#define WHITE		1
#define RED		2
#define GREEN		3
#define BLUE		4
#define YELLOW		5
#define CYAN		6
#define GREY		7

/* external prototypes */
int GetStringDisplayLength(char *str,int font_size);  /* char_width.c */



/* local prototypes */
int MakeSymbol(FILE *fp, unsigned int pins, int inputBubbles, 
	       int outputBubbles,
	       int (*body)(FILE *, int, int, unsigned int, unsigned int));

int AndBody(FILE *fp, int x, int y, unsigned int pins, unsigned int colour);
int OrBody(FILE *fp, int x, int y, unsigned int pins, unsigned int colour);
int XorBody(FILE *fp, int x, int y, unsigned int pins, unsigned int colour);
int Pin(FILE *fp, int x1, int y1, int x2, int y2, int bubble);
int PinAttribute(FILE *fp, int x, int y, unsigned int n, char *value); 
int WidenBody(FILE *fp, int x, int y, unsigned int pins, unsigned int colour);

/* globals */
unsigned int PinSpacing = 200;
unsigned int PinLength  = 300;

struct Table
{
  char         *name;     /* base name of part */
  unsigned int suffix;    /* suffix for part, 1 for normal, 2 for deMorgan.. */
  int (*body)(FILE *, int, int, unsigned int, unsigned int);  /* body style */
  int          inputBubbles;  /* where to draw bubbles */
  int          outputBubbles;
};


struct Table generate[] = {
  {"and",  1, AndBody, 0, 0},
  {"and",  2, OrBody,  1, 1},
  {"nand", 1, AndBody, 0, 1},
  {"nand", 2, OrBody,  1, 0},
  {"or",   1, OrBody,  0, 0},
  {"or",   2, AndBody, 1, 1},
  {"nor",  1, OrBody,  0, 1},
  {"nor",  2, AndBody, 1, 0},
  {"xor",  1, XorBody, 0, 0},
  {"xnor", 1, XorBody, 0, 1}
};

unsigned int nGenerate = sizeof(generate)/sizeof(struct Table);

int
main(int argc, char **argv)
{
  int i,j;
  int rc;
  char name[127];

  FILE *fp;

  printf("Making verilog symbols\n");

  for(i=0; i<nGenerate; i++) /* loop over table entries */
    for(j=2; j<10; j++)      /* loop over number of pins */
      {
	/* build file name */
	sprintf(name, "sym/%s%u-%u.sym",
		generate[i].name, j, generate[i].suffix);
	printf("Working on:%s\n",name);
	fp = fopen(name, "wb");
	if(fp == NULL)
	  {
	    fprintf(stderr,"Error: Unable to create file `%s' in %s()\n",
		    name, __func__);
	    return 1;
	  }

	rc = MakeSymbol(fp, j, 
			generate[i].inputBubbles, generate[i].outputBubbles,
			generate[i].body);
	if(rc)
	  {
	    fprintf(stderr,"Error: Symbol creation failed in %s()\n",
		    __func__);
	    return 1;
	  }

	/* and finally add the device attribute */
	fprintf(fp,"T 400 100 5 8 0 0 0 0\ndevice=%s\n",generate[i].name);
	/* and the positional pin directive */
	fprintf(fp,"T 400 200 5 8 0 0 0 0\nVERILOG_PORTS=POSITIONAL");
	fclose(fp);
      }

  return 0;
}

/* output a complete symbol having the desired body
 * and requested number of pins.  Draw bubbles on the input and/or
 * output pins.
 */
int
MakeSymbol(FILE *fp, unsigned int pins, int inputBubbles, int outputBubbles,
	   int (*body)(FILE *, int, int, unsigned int, unsigned int))
{
  unsigned int i;
  int rc;

  int bodyx, bodyy;       /* origin of body */
  int outputx, outputy;   /* origin of output pin */
  int firstx, firsty;     /* first end of first input pin */
  int totalHeight;

  int pinx, piny;

  int bodyWidth  = 700;
  int bodyHeight = 600;
  unsigned int pinCount = 1;

  char pinName[20];       /* temp for pinnname */

  if(fp == NULL)
    {
      fprintf(stderr, "Error: NULL file pointer passed to %s()\n",
	      __func__);
      return 1;
    }

  if(body == NULL)
    {
      fprintf(stderr, "Error: NULL body drawing function pointer passed "
	      "to %s()\n",
	      __func__);
      return 1;
    }

  /* do special pin spacing for 2 pins */
  if(pins == 2)
    PinSpacing = 400;
  else
    PinSpacing = 200;


  /* first, compute locations needed */
  totalHeight = bodyHeight;
  if(pins >= 3) /* do we need to add to the height ? */
    totalHeight += (pins-3) * PinSpacing;

  
  bodyx = PinLength;
  bodyy = totalHeight/2;

  outputx = bodyx+bodyWidth;
  outputy = bodyy;

  firstx  = PinLength;
  firsty  = 100;

  /* draw the body */
  rc = (*body)(fp, bodyx, bodyy, pins, GREEN);
  if(rc)
    {
      fprintf(stderr, "Error: Body function failed in %s()\n",
	      __func__);
      return 1;
    }

  /* draw the pins and attach appropriate attributes */
  rc = Pin(fp, outputx, outputy, outputx+PinLength, outputy, outputBubbles);
  if(rc)
    {
      fprintf(stderr, "Error: Pin drawing function failed in %s() "
	      "for output pin\n",
	      __func__);
      return 1;
    }
  /* attach pin attribute */
  rc = PinAttribute(fp, outputx, outputy, pinCount++, "OUT"); 
  if(rc)
    {
      fprintf(stderr, "Error: Pin Attribute function failed for output pin "
	      "in %s()\n",__func__);
      return 1;
    }

  /* do input pins */
  for(i=0; i<pins; i++)
    {
      /* calculate the position of the pin */
      pinx = firstx;
      piny = firsty+i*PinSpacing;
      /* output a pin */
      rc = Pin(fp, pinx, piny, pinx-PinLength, piny, inputBubbles);
      if(rc)
	{
	  fprintf(stderr,"Error: Pin drawing function failed for pin %u "
		  "in %s()\n",
		  i, __func__);
	  return 1;
	}
      /* output the attributes */
      sprintf(pinName,"IN%u",i);
      rc = PinAttribute(fp, pinx, piny, pinCount++, pinName);
      if(rc)
	{
	  fprintf(stderr,"Error: Pin Attributes function failed for pin %u "
		  "in %s()\n", i, __func__);
	  return 1;
	}

    }


  /* drop on a template uref attribute */
  fprintf(fp,"T %d %d 5 10 1 1 0 2\nrefdes=U?\n",
	  bodyx+100,
	  bodyy-400);


  return 0;
}




/* produce an and shaped body at the given offset.
 * The origin is along the horizontal midline of the shape.
 */
int 
AndBody(FILE *fp, int x, int y, unsigned int pins, unsigned int colour)
{

  if(fp == NULL)
    {
      fprintf(stderr, "Error: NULL file pointer passed to %s()\n",
	      __func__);
      return 1;
    }

  /* top and bottom lines */
  fprintf(fp, "L %d %d %d %d %u\n",
	  x, y+300, x+400, y+300, colour); 
  fprintf(fp, "L %d %d %d %d %u\n",
	  x, y-300, x+400, y-300, colour);
 
  /* left line */
  fprintf(fp, "L %d %d %d %d %u\n",
	  x, y-300, x, y+300, colour);

  /* arc at right */
  fprintf(fp, "A %d %d %d %d %d %u\n",
	  x+400, y, 300, -90, 180, colour);

  return WidenBody(fp, x, y, pins, colour);
}

/* produce a or shaped body at the given offset.
 * The origin is along the horizontal midline of the shape.
 */
int OrBody(FILE *fp, int x, int y, unsigned int pins, unsigned int colour)
{
  if(fp == NULL)
    {
      fprintf(stderr, "Error: NULL file pointer passed to %s()\n",
	      __func__);
      return 1;
    }

  /* top and bottom lines */
  fprintf(fp, "L %d %d %d %d %u\n",
	  x, y+300, x+300, y+300, colour);
  fprintf(fp, "L %d %d %d %d %u\n",
	  x, y-300, x+300, y-300, colour);

  /* left arc */
  fprintf(fp, "A %d %d %d %d %d %u\n",
	  x-260, y, 400, -48, 97, colour); 

  /* right top and bottom arcs */
  fprintf(fp, "A %d %d %d %d %d %u\n",
	  x+300, y+100, 400, 270, 76, colour); 
  fprintf(fp, "A %d %d %d %d %d %u\n",
	  x+300, y-100, 400, 90, -76, colour);
 
  return WidenBody(fp, x, y, pins, colour);
}

/* produce a xor shaped body at the given offset.
 * The origin is along the horizontal midline of the shape.
 */
int XorBody(FILE *fp, int x, int y, unsigned int pins, unsigned int colour)
{
  if(fp == NULL)
    {
      fprintf(stderr, "Error: NULL file pointer passed to %s()\n",
	      __func__);
      return 1;
    }

  /* top and bottom lines */
  fprintf(fp, "L %d %d %d %d %u\n",
	  x+100, y+300, x+300, y+300, colour);
  fprintf(fp, "L %d %d %d %d %u\n",
	  x+100, y-300, x+300, y-300, colour);

  /* left arc 1 */
  fprintf(fp, "A %d %d %d %d %d %u\n",
	  x-260, y, 400, -48, 97, colour);  
 /* left arc 2 */
  fprintf(fp, "A %d %d %d %d %d %u\n",
	  x-160, y, 400, -48, 97, colour); 
  

  /* right top and bottom arcs */
  fprintf(fp, "A %d %d %d %d %d %u\n",
	  x+300, y+100, 400, 270, 76, colour); 
  fprintf(fp, "A %d %d %d %d %d %u\n",
	  x+300, y-100, 400, 90, -76, colour); 

  return WidenBody(fp, x, y, pins, colour);
}


/* output two line segments that will serve to `widen' the
 * body of one of the above gates
 */
int 
WidenBody(FILE *fp, int x, int y, unsigned int pins, unsigned int colour)
{
  unsigned int distanceNeeded;

  if(fp == NULL)
    {
      fprintf(stderr, "Error: NULL file pointer passed to %s()\n",
	      __func__);
      return 1;
    }

  /* only need to do work if the number of pins > 2 */
  if(pins <= 3)
    return 0;


  /* Compute the amount needed to add to both top and bottom */
  distanceNeeded = (pins-3) * PinSpacing/2;

  /* output the line segments */
  /* for the top */
  fprintf(fp, "L %d %d %d %d %u\n",
	  x, y+300, x, y+300+distanceNeeded, colour);
  /* for the bottom */
  fprintf(fp, "L %d %d %d %d %u\n",
	  x, y-300, x, y-300-distanceNeeded, colour);

  return 0;
}


/* Draw a pin, optionally with a bubble
 */
int 
Pin(FILE *fp, int x1, int y1, int x2, int y2, int bubble) 
{
  int bx, by;              /* bubble x and y locations */
  int px1, py1, px2, py2;  /* pin x and y locations */
  int dx, dy;
  int br = 50;             /* bubble radius */

  double x,y;              /* unit vector in direction of line */
  double denom;

  if(fp == NULL)
    {
      fprintf(stderr, "Error: NULL file pointer passed to %s()\n",
	      __func__);
      return 1;
    }

  /* figure where pin ends are */
  if(bubble) /* is there a bubble? */
    { /* yes there is a bubble */

      /* compute a unit vector */
      dx = x2-x1;
      dy = y2-y1;
      denom = sqrt( dx*dx + dy*dy);
      if(denom < 1e-6)
	{
	  fprintf(stderr, "Error: Length of pin too small in %s()\n",
		  __func__);
	  return 1;
	}
      
      x = dx / denom;
      y = dy / denom;

      /* figure center of bubble */
      bx = x1 + x * br;
      by = y1 + y * br;

      /* figure location of first line end */
      px1 = x1 + x * 2 * br;
      py1 = y1 + y * 2 * br;
      
      px2 = x2;
      py2 = y2;
      

      /* draw the bubble */
      fprintf(fp, "V %d %d %d %d\n",
	      (int) bx, (int) by, br, CYAN);

    }
  else
    { /* no bubble, nothing special to do */

      px1 = x1; py1 = y1;
      px2 = x2; py2 = y2;

    }

  /* draw the pin */
  fprintf(fp, "P %d %d %d %d %d\n",
	  px1, py1, px2, py2, WHITE);

  return 0;
}

int 
PinAttribute(FILE *fp, int x, int y, unsigned int n, char *value) 
{
  if(fp == NULL)
    {
      fprintf(stderr, "Error: NULL file pointer passed to %s()\n",
	      __func__);
      return 1;
    }
  
  if(value == NULL)
    {
      fprintf(stderr, "Error: NULL value pointer passed to %s()\n",
	      __func__);
      return 1;
    }

  fprintf(fp, "{\n");
  fprintf(fp, "T %d %d %u %u 0 0 0 0\n",x, y, YELLOW, 8);
  fprintf(fp, "pin%u=%s\n",n, value);
  fprintf(fp, "}\n");
  
  return 0;
}
